pkg/collect/remote_collector.go

```diff
@@ -0,0 +1,372 @@
+			FileSize:                    c.Collect.FilesystemPerformance.FileSize,
+			Sync:                        c.Collect.FilesystemPerformance.Sync,
+			Datasync:                    c.Collect.FilesystemPerformance.Datasync,
+			Timeout:                     c.Collect.FilesystemPerformance.Timeout,
+			EnableBackgroundIOPS:        c.Collect.FilesystemPerformance.EnableBackgroundIOPS,
+			BackgroundIOPSWarmupSeconds: c.Collect.FilesystemPerformance.BackgroundIOPSWarmupSeconds,
+			BackgroundWriteIOPS:         c.Collect.FilesystemPerformance.BackgroundWriteIOPS,
+			BackgroundReadIOPS:          c.Collect.FilesystemPerformance.BackgroundReadIOPS,
+			BackgroundWriteIOPSJobs:     c.Collect.FilesystemPerformance.BackgroundWriteIOPSJobs,
+			BackgroundReadIOPSJobs:      c.Collect.FilesystemPerformance.BackgroundReadIOPSJobs,
+		}
+	} else if c.Collect.Certificate != nil {
+		localCollector.Certificate = &troubleshootv1beta2.Certificate{
+			HostCollectorMeta: troubleshootv1beta2.HostCollectorMeta{
+				CollectorName: c.Collect.Certificate.CollectorName,
+				Exclude:       c.Collect.Certificate.Exclude,
+			},
+			CertificatePath: c.Collect.Certificate.CertificatePath,
+			KeyPath:         c.Collect.Certificate.KeyPath,
+		}
+	} else if c.Collect.HostServices != nil {
+		localCollector.HostServices = &troubleshootv1beta2.HostServices{
+			HostCollectorMeta: troubleshootv1beta2.HostCollectorMeta{
+				CollectorName: c.Collect.HostServices.CollectorName,
+				Exclude:       c.Collect.HostServices.Exclude,
+			},
+		}
+	} else {
+		return nil, errors.New("no spec found to run")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
+	defer cancel()
+
+	result, err := c.RunRemote(ctx, localCollector, remoteCollectorNamePrefix, remoteCollectorDefaultInterval)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to run collector remotely")
+	}
+
+	if c.Redact {
+		result, err = redactMap(result, globalRedactors)
+	}
+
+	return result, nil
+}
+
+func (c *RemoteCollector) RunRemote(ctx context.Context, collector *troubleshootv1beta2.HostCollect, namePrefix string, interval time.Duration) (map[string][]byte, error) {
+	client, err := kubernetes.NewForConfig(c.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	scheme := runtime.NewScheme()
+	if err := corev1.AddToScheme(scheme); err != nil {
+		return nil, errors.Wrap(err, "failed to add runtime scheme")
+	}
+
+	// Get all the nodes where a Pod should be running
+	nodes, err := listNodesInSelectors(ctx, client, c.LabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the list of nodes matching a nodeSelector, got err: %w", err)
+	}
+
+	nameGenerator := names.SimpleNameGenerator
+
+	results := make(map[string][]byte)
+	errs := []error{}
+
+	resCh := make(chan map[string][]byte)
+	errCh := make(chan error)
+	waitCh := make(chan struct{})
+	doneCh := make(chan struct{})
+
+	go func() {
```

Comment:
Thanks, errgroup is a much simpler implementation.  I'm still a little unsure about the failure case though.  If a remote collector fails then with the current implementation it will cancel the other collectors and return the error.  

It might be better to allow the remaining nodes to continue and give a warning that one or more nodes failed.  I think it depends on the use case, so leaving it "strict" for now so that failures aren't hidden.   Now that retries have been added this should be less of an issue.

I've left adding `-race` to `make test` for a separate PR, but I can confirm it passes.